pkg/controller/volume/expand/pvc_populator.go

```diff
@@ -39,27 +52,39 @@ type PVCPopulator interface {
 }
 
 type pvcPopulator struct {
-	loopPeriod time.Duration
-	resizeMap  cache.VolumeResizeMap
-	pvcLister  corelisters.PersistentVolumeClaimLister
-	pvLister   corelisters.PersistentVolumeLister
-	kubeClient clientset.Interface
+	loopPeriod      time.Duration
+	resizeMap       cache.VolumeResizeMap
+	pvcLister       corelisters.PersistentVolumeClaimLister
+	pvLister        corelisters.PersistentVolumeLister
+	kubeClient      clientset.Interface
+	volumePluginMgr volume.VolumePluginMgr
+	cloud           cloudprovider.Interface
+	recorder        record.EventRecorder
 }
 
 func NewPVCPopulator(
 	loopPeriod time.Duration,
 	resizeMap cache.VolumeResizeMap,
 	pvcLister corelisters.PersistentVolumeClaimLister,
 	pvLister corelisters.PersistentVolumeLister,
-	kubeClient clientset.Interface) PVCPopulator {
+	plugins []volume.VolumePlugin,
+	cloud cloudprovider.Interface,
+	kubeClient clientset.Interface) (PVCPopulator, error) {
 	populator := &pvcPopulator{
 		loopPeriod: loopPeriod,
 		pvcLister:  pvcLister,
 		pvLister:   pvLister,
 		resizeMap:  resizeMap,
+		cloud:      cloud,
 		kubeClient: kubeClient,
 	}
-	return populator
+	if err := populator.volumePluginMgr.InitPlugins(plugins, nil, populator); err != nil {
```

Comment:
Is there any advantage over initializing plugin manager here again rather than just passing the one that already has been initialzied in expand_controller ?

Reply:
@gnufied govet doesn't like passing `volumePluginMgr` as an argument (see the first commit in my PR) because it results in passing a lock by value:
```
2018/07/30 19:02:56 Code for OpenAPI definitions generated
# k8s.io/kubernetes/pkg/controller/volume/expand
pkg/controller/volume/expand/expand_controller.go:145: call of NewPVCPopulator copies lock value: k8s.io/kubernetes/pkg/volume.VolumePluginMgr contains sync.Mutex
pkg/controller/volume/expand/pvc_populator.go:64: NewPVCPopulator passes lock by value: k8s.io/kubernetes/pkg/volume.VolumePluginMgr contains sync.Mutex
pkg/controller/volume/expand/pvc_populator.go:71: literal copies lock value from volumePluginMgr: k8s.io/kubernetes/pkg/volume.VolumePluginMgr contains sync.Mutex
make[1]: *** [vet] Error 1
```
PVC populator implementing the `VolumeHost` interface is precisely to handle this situation. As I said, I'm not a fan of code duplication, but this is a side-effect of the original implementation where we have two controllers more or less doing the same job.

Reply:
Can this not be worked around by passing the volume plugin manager as a pointer rather than by value?